File: fx/src/main/java/gargoyle/fx/log/FXLog.java
package gargoyle.fx.log;

import java.util.Map;

@SuppressWarnings({"unused", "ClassWithTooManyMethods"})
public final class FXLog {

    private FXLog() {
        throw new IllegalStateException(FXLog.class.getName());
    }

    public static void debug(final String msg) {
        log(Level.DEBUG, msg);
    }

    public static void debug(final String format, final Map<String, ?> arguments) {
        log(Level.DEBUG, format, arguments);
    }

    public static void debug(final Throwable t, final String msg) {
        log(t, Level.DEBUG, msg);
    }

    public static void debug(final Throwable t, final String msg, final Map<String, ?> arguments) {
        log(t, Level.DEBUG, msg, arguments);
    }

    public static void error(final String msg) {
        log(Level.ERROR, msg);
    }

    public static void error(final String format, final Map<String, ?> arguments) {
        log(Level.ERROR, format, arguments);
    }

    public static void error(final Throwable t, final String msg) {
        log(t, Level.ERROR, msg);
    }

    public static void error(final Throwable t, final String msg, final Map<String, ?> arguments) {
        log(t, Level.ERROR, msg, arguments);
    }

    public static void info(final String msg) {
        log(Level.INFO, msg);
    }

    public static void info(final String format, final Map<String, ?> arguments) {
        log(Level.INFO, format, arguments);
    }

    public static void info(final Throwable t, final String msg) {
        log(t, Level.INFO, msg);
    }

    public static void info(final Throwable t, final String msg, final Map<String, ?> arguments) {
        FXLogUtil.log(t, Level.INFO, msg, arguments);
    }

    public static void log(final Level level, final String msg) {
        FXLogUtil.log(null, level, msg, Map.of());
    }

    public static void log(final Throwable t, final Level level, final String msg) {
        FXLogUtil.log(t, level, msg, Map.of());
    }

    public static void log(final Level level, final String format, final Map<String, ?> arguments) {
        FXLogUtil.log(null, level, format, arguments);
    }

    public static void log(final Throwable t, final Level level, final String format, final Map<String, ?> arguments) {
        FXLogUtil.log(t, level, format, arguments);
    }

    public static void trace(final String msg) {
        log(Level.TRACE, msg);
    }

    public static void trace(final String format, final Map<String, ?> arguments) {
        log(Level.TRACE, format, arguments);
    }

    public static void trace(final Throwable t, final String msg) {
        log(t, Level.TRACE, msg);
    }

    public static void trace(final Throwable t, final String msg, final Map<String, ?> arguments) {
        log(t, Level.TRACE, msg, arguments);
    }

    public static void warn(final String msg) {
        log(Level.WARN, msg);
    }

    public static void warn(final String format, final Map<String, ?> arguments) {
        log(Level.WARN, format, arguments);
    }

    public static void warn(final Throwable t, final String msg) {
        log(t, Level.WARN, msg);
    }

    public static void warn(final Throwable t, final String msg, final Map<String, ?> arguments) {
        log(t, Level.WARN, msg, arguments);
    }

}

File: fx/src/main/kotlin/gargoyle/fx/log/FxLog.kt
package gargoyle.fx.log

object FxLog {

    fun error(message: String): Unit =
        log(FxLogLevel.ERROR, message, null)

    fun error(message: String, t: Throwable? = null): Unit =
        log(FxLogLevel.ERROR, message, t)

    fun error(message: () -> String): Unit =
        log(FxLogLevel.ERROR, message, null)

    fun error(message: () -> String, t: Throwable? = null): Unit =
        log(FxLogLevel.ERROR, message, t)

    fun warn(message: String): Unit =
        log(FxLogLevel.WARN, message, null)

    fun warn(message: String, t: Throwable? = null): Unit =
        log(FxLogLevel.WARN, message, t)

    fun warn(message: () -> String): Unit =
        log(FxLogLevel.WARN, message, null)

    fun warn(message: () -> String, t: Throwable? = null): Unit =
        log(FxLogLevel.WARN, message, t)

    fun info(message: String): Unit =
        log(FxLogLevel.INFO, message, null)

    fun info(message: String, t: Throwable? = null): Unit =
        log(FxLogLevel.INFO, message, t)

    fun info(message: () -> String): Unit =
        log(FxLogLevel.INFO, message, null)

    fun info(message: () -> String, t: Throwable? = null): Unit =
        log(FxLogLevel.INFO, message, t)

    fun debug(message: String): Unit =
        log(FxLogLevel.DEBUG, message, null)

    fun debug(message: String, t: Throwable? = null): Unit =
        log(FxLogLevel.DEBUG, message, t)

    fun debug(message: () -> String): Unit =
        log(FxLogLevel.DEBUG, message, null)

    fun debug(message: () -> String, t: Throwable? = null): Unit =
        log(FxLogLevel.DEBUG, message, t)

    fun trace(message: String): Unit =
        log(FxLogLevel.TRACE, message, null)

    fun trace(message: String, t: Throwable? = null): Unit =
        log(FxLogLevel.TRACE, message, t)

    fun trace(message: () -> String): Unit =
        log(FxLogLevel.TRACE, message, null)

    fun trace(message: () -> String, t: Throwable? = null): Unit =
        log(FxLogLevel.TRACE, message, t)

    fun log(level: FxLogLevel, message: String): Unit =
        FxLoggerFactory.logger.log(stackTraceElement, level, { message }, null)

    fun log(level: FxLogLevel, message: String, t: Throwable? = null): Unit =
        FxLoggerFactory.logger.log(stackTraceElement, level, { message }, t)

    fun log(level: FxLogLevel, message: () -> String): Unit =
        FxLoggerFactory.logger.log(stackTraceElement, level, message, null)

    fun log(level: FxLogLevel, message: () -> String, t: Throwable? = null): Unit =
        FxLoggerFactory.logger.log(stackTraceElement, level, message, t)

    private val stackTraceElement: StackTraceElement?
        get() = Thread.currentThread().stackTrace.firstOrNull { element: StackTraceElement ->
            arrayOf("java.", "javax.", FxLog::class.java.packageName).none { element.className.startsWith(it) }
        }
}
